Document certificate authentication (#331)
timescale/docs.timescale.com-content: getting-started/data-node-authentication.md

# Data Node Authentication [](distdb-auth)

In a [multi-node environment][scaling-out], it is necessary to ensure
that the access node have the necessary roles and permissions on all

> the data nodes where distributed hypertables are placed.

Reviewer: ```suggestion
that the access node has the necessary roles and permissions on all
data nodes where distributed hypertables are placed.
```

key. For the access node, the root certificate can be shared and we
only need a single root certificate on the access node.

For the data node, we need a certificate and key specially generated
for the data node. The root certificate can be copied from the access
node.

### Setting up a Certification Authority

A Certificate Authority is needed as a trusted third party, so you
need to be able to sign other certificates. If you are in a production
environment, you probably have a CA already setup, in which case you
can use that instead. If you do not have a CA key set up, this section
is for you.

The key of the certificate authority is used to sign Certificate
Signing Requests (CSRs), and the certificate of the certificate
authority is what is used as a root certificate for other parties.

A common way to set up certificate authentication when demonstrating
it is using self-signed certificates for the clients, but in this
chapter you will instead use a toy certificate authority. This is more
similar to how it will work in a real production setup.

>:WARNING: The Certificate Authority key is central to your security

> >and need to be treated with care. Ideally, the key should be

Reviewer: ```suggestion
>and needs to be treated with care. Ideally, the key should be
```

```bash
openssl genpkey -algorithm rsa -out auth.key
```

To generate a root certificate for the toy CA, you can generate a
self-signed root certificate using `openssl`.

```
$ openssl req -new -key auth.key -days 3650 -out root.crt -x509
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [AU]:US
State or Province Name (full name) [Some-State]:New York
Locality Name (eg, city) []:New York
Organization Name (eg, company) [Internet Widgits Pty Ltd]:Example Company Pty Ltd
Organizational Unit Name (eg, section) []:
Common Name (e.g. server FQDN or YOUR name) []:http://cert.example.com/
Email Address []: 
```

The common name can be anything for this certificate, but here the
common name contains the (imaginary) location of the root certificate,
which in theory would allow a user to download the certificate from a

> know location and verify that it is the correct one.

Reviewer: ```suggestion
known location and verify that it is the correct one.
```

> Remember that the root certificate only contain public information and

Reviewer: ```suggestion
Remember that the root certificate only contains public information and
```

> can be distributed freely between machine. Actually, it is more secure

Reviewer: ```suggestion
can be distributed freely between machines. Actually, it is more secure
```

to distribute it widely since anybody interested in using the root

> certificate can be more sure that she got the right certificate if it

Reviewer: ```suggestion
certificate can be more certain that she got the right certificate if it
```

is available at many locations.

> ### Generating key and certificate for the instances

Reviewer: ```suggestion
### Generating keys and certificates for the instances
```

   directory of the PostgreSQL instance, if they are not already
   there.

4. Copy the root certificate file `root.crt` from the certificate
   authority into the data directory of the instance.

### Configure the instance to use SSL authentication

You now need to configure the instance to use SSL authentication by
setting the `ssl` option to `on` in the `postgresql.conf`
configuration file and set `ssl_ca_file` in the configuration file for
the instance.

Typically this is done on the data node, but you might want to do that
for the access node as well if you want to use certificate
authentication to log into the access node as well.

```
ssl = on
ssl_ca_file = 'root.crt'
ssl_cert_file = 'server.crt'
ssl_key_file = 'server.key'
```

Strictly speaking, you do not need to set `ssl_cert_file` and
`ssl_key_file` since you are using the default file names here, but it
does not hurt to set them explicitly.

After you have edited the `postgresql.conf` on the data node, you need
to configure `pg_hba.conf` to accept certificates for users, in
particular, you want the access node to be accepted provided it has a
properly signed certificate. If you add the line below, any user that

> belong to the role `ssl_cert` will be allowed to log in.

Reviewer: ```suggestion
belongs to the role `ssl_cert` will be allowed to log in.
```

```
hostssl all +ssl_cert all cert clientcert=1
```

Also create the role `ssl_cert` on the data node.

```sql
CREATE ROLE ssl_cert;
```

### Generate user certificates

The access node does not have any user certificates, so it cannot yet
log into the data node. The user certificates are normally stored in
`timescaledb/certs` under the data directory, so we need to store the
user key files and user certificate files here.

>:TIP: The location of the user certificates and keys are relative to
>`ssl_dir`, so you can place it outside the data directory by setting
>`ssl_dir` to some other value.

The key and certificate file names are constructed by taking the MD5

> sum of the role name and use it as the base for both the key and the

Reviewer: ```suggestion
sum of the role name and using it as the base for both the key and the
```

certificate files. This means that the key file for `postgres` will be
`e8a48653851e28c69d0506508fb27fc5.key` and the certificate file name
will be `e8a48653851e28c69d0506508fb27fc5.crt`.

We generate a key and certificate file in a similar way to how we do
it for the instance key and certificate, but sign it using the
instance key instead of the certification authority. In the data

> directory of the access node, you can do in in the following steps:

Reviewer: ```suggestion
directory of the access node, you can do it through the following steps:
```

1. Compute the base name for the files, generate a subject
   identifier, and create names for the key and certificate files. 

   ```bash
   pguser=postgres
   base=`echo -n $pguser | md5sum | cut -c1-32`
   subj=/C=US/ST=New York/L=New York/O=Timescale/OU=Engineering/CN=$pguser
   key_file=timescaledb/certs/$base.key
   crt_file=timescaledb/certs/$base.crt
   ```

   Here we copy most of the data from the server certificate for the

>    subject, but the common name (`CN`) need to be set to the user

Reviewer: ```suggestion
   subject, but the common name (`CN`) needs to be set to the user
```

  and should be kept secure on the instance where it is generated.

For each user on the access node, we need a signed certificate and a
key. For the access node, the root certificate can be shared and we
only need a single root certificate on the access node.

For the data node, we need a certificate and key specially generated
for the data node. The root certificate can be copied from the access
node.

### Setting up a Certification Authority

A Certificate Authority is needed as a trusted third party, so you
need to be able to sign other certificates. If you are in a production
environment, you probably have a CA already setup, in which case you
can use that instead. If you do not have a CA key set up, this section
is for you.

The key of the certificate authority is used to sign Certificate
Signing Requests (CSRs), and the certificate of the certificate
authority is what is used as a root certificate for other parties.

A common way to set up certificate authentication when demonstrating
it is using self-signed certificates for the clients, but in this
chapter you will instead use a toy certificate authority. This is more

> similar to how it will work in a real production setup.

Reviewer: ```suggestion
similar to a real production setup.
```

Common Name (e.g. server FQDN or YOUR name) []:http://cert.example.com/
Email Address []: 
```

The common name can be anything for this certificate, but here the
common name contains the (imaginary) location of the root certificate,
which in theory would allow a user to download the certificate from a
known location and verify that it is the correct one.

Remember that the root certificate only contains public information and
can be distributed freely between machines. Actually, it is more secure
to distribute it widely since anybody interested in using the root
certificate can be more certain that she got the right certificate if it
is available at many locations.

### Generating keys and certificates for the instances

Once you have a root certificate authority you can create certificates
for the both the data node and the access node. On the data node you
need a signed certificate to certify to the access node that the data
node is valid. On the access node you need to have a signed
certificate so that the access node can sign user certificates.

The default names for the instance key and certificate is `server.key`
and `server.crt` respectively and they are placed in the data

> directory of the instance, so let's use those names. If you do not

Reviewer 2: I remember we had a PR which purpose was to place those files in a user/certificate directory inside the data directory. I had an impression that we merged it or we just not expose those details here?
